File: runtime/org.eclipse.etrice.runtime.cpp/src/common/modelbase/InterfaceItemBase.cpp
/*******************************************************************************
 * Copyright (c) 2012 Draeger Medical GmbH (http://www.draeger.com).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * CONTRIBUTORS:
 * 		Peter Karlitschek (initial contribution)
 *
 *******************************************************************************/

#include "common/messaging/MessageServiceController.h"
#include "common/messaging/RTObject.h"
#include "common/messaging/RTServices.h"
#include "common/modelbase/IEventReceiver.h"
#include "common/modelbase/IInterfaceItemOwner.h"
#include "common/modelbase/InterfaceItemBase.h"
#include "common/modelbase/IReplicatedInterfaceItem.h"

namespace etRuntime {

void InterfaceItemBase::connect(IRTObject* obj, const String& path1, const String& path2) {
	IRTObject* obj1 = obj->getObject(path1);
	IRTObject* obj2 = obj->getObject(path2);

	IInterfaceItem* ifItem1 = dynamic_cast<IInterfaceItem*>(obj1);
	IInterfaceItem* ifItem2 = dynamic_cast<IInterfaceItem*>(obj2);

	if (ifItem1 != 0 && ifItem2 != 0) {
		ifItem1->connectWith(ifItem2);
	}
}

InterfaceItemBase::InterfaceItemBase(IInterfaceItemOwner* owner, const String& name, int localId, int idx) :
		AbstractMessageReceiver(owner->getEventReceiver(), name),
		m_localId(localId),
		m_idx(idx),
		m_peerAddress(Address::EMPTY),
		m_peer(0),
		m_ownMsgReceiver(0),
		m_peerMsgReceiver(0),
		m_replicator(0) {

	m_replicator = dynamic_cast<IReplicatedInterfaceItem*>(owner);
}

InterfaceItemBase::~InterfaceItemBase() {
	m_peerAddress = Address::EMPTY;
	m_ownMsgReceiver = 0;
	m_peerMsgReceiver = 0;
}

void InterfaceItemBase::init() {
	int thread = dynamic_cast<IEventReceiver*> (getParent())->getThread();
	if (thread >= 0) {
		IMessageService* msgSvc = RTServices::getInstance().getMsgSvcCtrl().getMsgSvc(thread);
		Address addr = msgSvc->getFreeAddress();
		setAddress(addr);
		msgSvc->addMessageReceiver(*this);

		m_ownMsgReceiver = msgSvc;
	}
}

IInterfaceItem* InterfaceItemBase::connectWith(IInterfaceItem* peer) {
	if (peer != 0) {
		m_peer = peer;

//			if (peer instanceof IInterfaceItemBroker) {
//				this.peer = peer.connectWith(this);
//				return this.peer;
//			}

		IReplicatedInterfaceItem* replPeer = dynamic_cast<IReplicatedInterfaceItem*>(m_peer);
		if (replPeer != 0) {
			m_peer = replPeer->createSubInterfaceItem();
		}

		InterfaceItemBase* ifItemPeer = dynamic_cast<InterfaceItemBase*>(m_peer);
		if (ifItemPeer != 0) {
			// initialize Ports
			this->init();
			ifItemPeer->init();

			// connect with each other
			m_peerAddress = ifItemPeer->getAddress();
			ifItemPeer->m_peerAddress = getAddress();
			m_peerMsgReceiver = ifItemPeer->m_ownMsgReceiver;
			ifItemPeer->m_peerMsgReceiver = m_ownMsgReceiver;
		}

	}

	return peer;
}

void InterfaceItemBase::disconnect() {
	disconnectInternal();
	if (m_peer != 0) {
		InterfaceItemBase* peer = dynamic_cast<InterfaceItemBase*>(m_peer);
		if (peer != 0)
			peer->disconnectInternal();
		m_peer = 0;
	}
}

void InterfaceItemBase::disconnectInternal() {
	m_peerAddress = Address::EMPTY;
	m_peerMsgReceiver = 0;

	if (m_replicator != 0)
		destroy();
}

IEventReceiver* InterfaceItemBase::getActor() const {
	return dynamic_cast<IEventReceiver*>(getParent());
}

void InterfaceItemBase::destroy() {
	if (m_peerAddress.isValid()) {
		disconnect();
	}

	if (m_replicator != 0) {
		m_replicator->removeItem(*this);
	}

	if(m_ownMsgReceiver) {
		m_ownMsgReceiver->removeMessageReceiver(*this);
		m_ownMsgReceiver->freeAddress(getAddress());
	}

	AbstractMessageReceiver::destroy();
}

String InterfaceItemBase::toString() const {
	char buffer [256];
	const char* sub = (m_replicator != 0) ? "sub " : "";
	int pos = sprintf(buffer, "%sport %s %s ", sub, getName().c_str(), getAddress().toID().c_str());
	if (m_peerMsgReceiver == 0) {
		sprintf(buffer+pos, "UNCONNECTED");
	}
	else {
		const char* peerName = (m_peer != 0) ? m_peer->getName().c_str() : "?";
		sprintf(buffer + pos, " -> %s %s", peerName, m_peerAddress.toID().c_str());
	}

	return buffer;
}

} /* namespace etRuntime */

File: runtime/org.eclipse.etrice.runtime.java/src/org/eclipse/etrice/runtime/java/modelbase/InterfaceItemBase.java
/*******************************************************************************
 * Copyright (c) 2010 protos software gmbh (http://www.protos.de).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *******************************************************************************/

package org.eclipse.etrice.runtime.java.modelbase;

import java.util.List;

import org.eclipse.etrice.runtime.java.messaging.AbstractMessageReceiver;
import org.eclipse.etrice.runtime.java.messaging.Address;
import org.eclipse.etrice.runtime.java.messaging.IMessageReceiver;
import org.eclipse.etrice.runtime.java.messaging.IMessageService;
import org.eclipse.etrice.runtime.java.messaging.IRTObject;
import org.eclipse.etrice.runtime.java.messaging.RTServices;

/**
 * The abstract base class for actor class interface items like {@link PortBase}
 * and {@link InterfaceItemBroker}s.
 * 
 * @author Henrik Rentz-Reichert
 */
public abstract class InterfaceItemBase extends AbstractMessageReceiver {
	
	/**
	 * If this is part of an {@link IReplicatedInterfaceItem} then the
	 * owner of this item is stored in this field.
	 * It is used for sub ports of a replicated ports. When of of this is
	 * disconnected then it is also removed from its parent and destroyed.
	 */
	private IReplicatedInterfaceItem replicator = null;
	
	protected IMessageService ownMsgReceiver;
	protected IMessageReceiver peerMsgReceiver;
	private int localId;
	private int idx;
	protected Address peerAddress = null;
	private InterfaceItemBase peer;


	/**
	 * The constructor determines the thread of its {@link IEventReceiver}
	 * 
	 * @param owner
	 * @param name
	 * @param localId
	 * @param idx
	 */
	public InterfaceItemBase (IInterfaceItemOwner owner, String name, int localId, int idx) {
		super(owner.getEventReceiver(), name);
		
		this.localId = localId;
		this.idx = idx;
		
		if (owner instanceof IReplicatedInterfaceItem)
			replicator = (IReplicatedInterfaceItem) owner;
		
		int thread = owner.getEventReceiver().getThread();
		if (thread>=0) {
			IMessageService msgSvc = RTServices.getInstance().getMsgSvcCtrl().getMsgSvc(thread);
			Address addr = msgSvc.getFreeAddress();
			setAddress(addr);
			msgSvc.addMessageReceiver(this);
			
			this.ownMsgReceiver = msgSvc;
		}
		
		connectWithPeer();
	}

	protected void connectWithPeer() {
		List<String> peerPaths = getParent().getPeersForPath(getInstancePath());
		if (peerPaths!=null && !peerPaths.isEmpty()) {
			IRTObject object = getObject(peerPaths.get(0));
			InterfaceItemBase peer = null;
			if (object instanceof InterfaceItemBase) {
				peer = ((InterfaceItemBase) object);
			}
			else if (object instanceof IReplicatedInterfaceItem) {
				peer = ((IReplicatedInterfaceItem) object).createSubInterfaceItem();
			}
			connectWith(peer);
		}
	}

	protected synchronized InterfaceItemBase connectWith(InterfaceItemBase peer) {
		if (peer!=null) {
			this.peer = peer;
			
			if (peer instanceof IInterfaceItemBroker) {
				this.peer = peer.connectWith(this);
				return this.peer;
			}
			
			// connect with each other
			peerAddress = peer.getAddress();
			peer.peerAddress = getAddress();
			this.peerMsgReceiver = peer.ownMsgReceiver;
			peer.peerMsgReceiver = ownMsgReceiver;
			
		}
		
		return peer;
	}
	
	protected synchronized void disconnect() {
		disconnectInternal();
		if (peer!=null) {
			peer.disconnectInternal();
			peer = null;
		}
	}

	private void disconnectInternal() {
		peerAddress = null;
		peerMsgReceiver = null;
		
		if (replicator!=null)
			destroy();
	}
	
	protected IMessageReceiver getMsgReceiver() {
		return ownMsgReceiver;
	}

	protected synchronized Address getPeerAddress() {
		return peerAddress;
	}

	protected synchronized IMessageReceiver getPeerMsgReceiver() {
		return peerMsgReceiver;
	}
	
	public IEventReceiver getActor() {
		return (IEventReceiver) getParent();
	}

	public int getLocalId() {
		return localId;
	}

	public int getIdx() {
		return idx;
	}
	
	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.RTObject#destroy()
	 */
	@Override
	protected void destroy() {
		if (peerAddress!=null) {
			disconnect();
		}
		
		if (replicator!=null) {
			replicator.removeItem(this);
		}
		
		ownMsgReceiver.removeMessageReceiver(this);
		ownMsgReceiver.freeAddress(getAddress());
		
		super.destroy();
	}
	
	@Override
	public String toString() {
		return ((replicator!=null)?"sub ":"")+"port "+getName()+" "+getAddress()+" <-> "+getPeerAddress();
	}
}
